### PR TITLE
Add AutoSuggest Field highlighting

### PR DESCRIPTION
**Pull request by Mottie**
## Description

In our form 526, Veterans enter disability names in an autosuggest-enabled input. During research sessions, Veterans were confused by the seemingly random results. To help alleviate this problem, we will highlight portions of the suggested list that match the user input (see screenshot)

- Set `ui:options.highlightText` option to `true` to enable (enabled in this PR for form 526 add new disability input)
- Uses `<mark>` element to highlight (easy to modify it's style)

## Original issue(s)

https://github.com/department-of-veterans-affairs/va.gov-team/issues/30054

## Testing done

Added unit test

## Screenshots

<details><summary>Current state</summary>

Screen Shot 2021-09-16 at 2 27 18 PM</details>

<details><summary>Highlighted suggestions</summary>

Screen Shot 2021-09-16 at 2 24 31 PM</details>

<details><summary>Work  across words</summary>

Screen Shot 2021-09-16 at 2 24 16 PM</details>

## Acceptance criteria

- [x] Enable highlight using `ui:options`
- [x] Unit tests all passing

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

**Review by ashagross — APPROVED** [review]

File: src/platform/forms-system/src/js/fields/AutosuggestField.jsx [hunk 1]
```diff
@@ -208,6 +208,19 @@ export default class AutosuggestField extends React.Component {
     const { idSchema, formContext, formData, uiSchema, schema } = this.props;
     const id = idSchema.$id;
 
+    // wrap matching text in a <mark> element
+    const highlightText = uiSchema['ui:options']?.highlightText;
```
Comment (cvalarida) [inline src/platform/forms-system/src/js/fields/AutosuggestField.jsx:212]:
```suggestion
    const highlightText = uiSchema['ui:options']?.highlightText ?? true;
```

This seems like a solid improvement, so I'd recommend enabling this by default. Fields can manually disable this as desired.

File: src/platform/forms-system/src/js/fields/AutosuggestField.jsx [hunk 2]
```diff
@@ -208,6 +208,19 @@ export default class AutosuggestField extends React.Component {
...
+    const value = this.state.input;
+    const regexp = new RegExp(`(${value})`);
+    const highLightMatchingText = query => {
+      if (value.length > 2) {
+        return query
+          .split(regexp)
+          .map(str => (str === value ? <mark>{value}</mark> : str));
```
Comment (cvalarida) [inline src/platform/forms-system/src/js/fields/AutosuggestField.jsx:219]: How does this affect screen readers?

cc @noahgelman

**Review by cvalarida — COMMENTED** [review]
The code looks good. Before approving, I'd like to get input from
- @noahgelman to make sure the markup change won't negatively impact screenreader users
    - Also, there might be something about the color contrast of the yellow highlight vs the white background, but I'll leave it to the professionals to give their recommendations there :sweat_smile: 
- Natalie Hill to make sure the design is okay

**Comment by Mottie** [comment]
Made an update to ignore the case
Screen Shot 2021-09-17 at 2 06 22 PM

This is the HTML snippet from the dropdown:

```html
<div class="autosuggest-list" role="listbox">
  <div id="downshift-1-item-0" role="option" aria-selected="false" class="autosuggest-item">epi<mark>the</mark>lioma
  </div>
  <div id="downshift-1-item-1" role="option" aria-selected="false" class="autosuggest-item">endo<mark>the</mark>lioma
  </div>
  <!-- ... -->
  <div id="downshift-1-item-19" role="option" aria-selected="false" class="autosuggest-item">Adhesions - Neurological
    o<mark>the</mark>r System</div>
</div>
```

**Review by GnatalieH — APPROVED** [review]
I will document the autosuggest pattern and highlight color in the design system.

File: src/platform/forms-system/src/js/fields/AutosuggestField.jsx [hunk 3]
```diff
@@ -208,6 +208,19 @@ export default class AutosuggestField extends React.Component {
...
 
+    // wrap matching text in a <mark> element
+    const highlightText = uiSchema['ui:options']?.highlightText ?? true;
+    const value = this.state.input?.toLowerCase() || '';
+    const regexp = new RegExp(`(${value})`, 'i');
+    const highLightMatchingText = query => {
+      if (value.length > 2) {
+        return query
+          .split(regexp)
+          .map(str => (str.toLowerCase() === value ? <mark>{str}</mark> : str));
```
Comment (cvalarida) [inline src/platform/forms-system/src/js/fields/AutosuggestField.jsx:219]: Same question for @noahgelman as before: Does this pose a problem for screen readers?

**Comment by noahgelman** [comment]
> I changed the wrapper to a span, but I feel like `mark` should be supported/ignored by screen readers 🤷🏽

Valid opinion.

This is approved from an accessibility standpoint

**Comment by Mottie** [comment]
@bkjohnson Can you pick up the review on this PR please 😺

**Comment by GnatalieH** [comment]
Can someone remind me what the highlight color is? I want to document it somewhere in the design system. Thanks.

**Comment by Mottie** [comment]
@GnatalieH The original `<mark>` was set to `#ff0` (yellow) which isn't in the design system. So I changed it to a `<span>` and used `vads-u-background-color--gold` - which is a tertiary color from https://design.va.gov/utilities/background-color

File: src/platform/forms-system/src/js/fields/AutosuggestField.jsx [hunk 4]
```diff
@@ -208,6 +208,26 @@ export default class AutosuggestField extends React.Component {
     const { idSchema, formContext, formData, uiSchema, schema } = this.props;
     const id = idSchema.$id;
 
+    // wrap matching text in a <span> element
+    const highlightText = uiSchema['ui:options']?.highlightText ?? true;
+    const value = this.state.input?.toLowerCase() || '';
+    const regexp = new RegExp(`(${value})`, 'i');
```
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/AutosuggestField.jsx:214]: Can this be renamed `caseInsensitiveMatch` or something just to be a bit more explicit?

**Review by bkjohnson — COMMENTED** [review]
It looks like there's a failing test, but things look good other than that. One small rename request.

**Comment by GnatalieH** [comment]
> @GnatalieH The original `<mark>` was set to `#ff0` (yellow) which isn't in the design system. So I changed it to a `<span>` and used `vads-u-background-color--gold` - which is a tertiary color from https://design.va.gov/utilities/background-color. If you have a suggestion on a better color, I'd appreciate any feedback. Thanks!
> 
> Example of the difference:
> 
> Screen Shot 2021-11-01 at 9 42 59 AM

I think as long as the tertiary gold color from the design system passes checks for contrast, this works. Thanks!

**Review by bkjohnson — APPROVED** [review]

**Pull request merged**